### ISSUE TITLE
Unexpected conversion error on large dataset

### ISSUE THREAD

**Issue by victorlin** [duckdb/duckdb#2860]
## What happens?

I am trying to filter on a date column, stored as `STRING` but casted to `DATE` in a view for easy comparisons. Some dates are incomplete (e.g. `2020-03`) which are ignored from casting by a regex, but the view contains all rows via an `OUTER JOIN`.

In small datasets, incomplete dates are ignored as expected. However, in large datasets, it seems like the incomplete dates are casted, resulting in a `Conversion Error`.

### 1. Working example (small dataset)

Generate data file for import:

```sh
echo "id,date" >> test1.csv
i=0;
while [ $i -lt 100 ]; do echo "$i,2020-03-04" >> test1.csv; i=$(($i+1)); done
while [ $i -lt 200 ]; do echo "$i,2020" >> test1.csv; i=$(($i+1)); done
while [ $i -lt 300 ]; do echo "$i,2020-03" >> test1.csv; i=$(($i+1)); done
```

Create table and view, run query:

```sql
CREATE TABLE test1 AS SELECT * FROM read_csv_auto('test1.csv', header=True);
CREATE VIEW test1_view AS SELECT
    test1.id,
    r.date
FROM test1
LEFT OUTER JOIN (
    SELECT
        id,
        date::DATE AS date
    FROM test1
    WHERE regexp_matches(date, '^\d{4}\-(0?[1-9]|1[012])\-(0?[1-9]|[12][0-9]|3[01])$')
) r
ON (test1.id = r.id);

SELECT count(*) FROM test1_view WHERE date > '2020-03-01';
```

Output:

```
┌──────────────┐
│ count_star() │
├──────────────┤
│ 100          │
└──────────────┘
```

### 2. Bug example (large dataset)

This is the same as the working example, except with 100x the data in the same proportions. I'm not sure what the exact cutoff is for where things start to break.

Generate data file for import:

```sh
echo "id,date" >> test2.csv
i=0;
while [ $i -lt 10000 ]; do echo "$i,2020-03-04" >> test2.csv; i=$(($i+1)); done
while [ $i -lt 20000 ]; do echo "$i,2020" >> test2.csv; i=$(($i+1)); done
while [ $i -lt 30000 ]; do echo "$i,2020-03" >> test2.csv; i=$(($i+1)); done
```

Create table and view, run query:

```sql
CREATE TABLE test2 AS SELECT * FROM read_csv_auto('test2.csv', header=True);
CREATE VIEW test2_view AS SELECT
    test2.id,
    r.date
FROM test2
LEFT OUTER JOIN (
    SELECT
        id,
        date::DATE AS date
    FROM test2
    WHERE regexp_matches(date, '^\d{4}\-(0?[1-9]|1[012])\-(0?[1-9]|[12][0-9]|3[01])$')
) r
ON (test2.id = r.id);

SELECT count(*) FROM test2_view WHERE date > '2020-03-01';
```

Output:

```
Error: Conversion Error: Conversion Error: date field value out of range: "2020-03", expected format is (YYYY-MM-DD)
```

Based on the `WHERE regexp_matches(...)`, the value `"2020-03"` shouldn't be casted. However, the output makes it look like a failed casting attempt.

Interestingly, it works fine when replacing `CREATE VIEW` with `CREATE TABLE`.

#### Environment:
 - OS: macOS 11.6.1
 - DuckDB Version: 0.3.1
 - DuckDB Client: CLI, Python

#### Before Submitting

- [X] **Have you tried this on the latest `master` branch?**
- [X] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by Mytherin** [comment]
Thanks for the report! This appears to be an issue with an optimizer. I will have a look.

**Comment by omo** [comment]
This might worth another look since the fix for #5342 might have fixed this too.

**Comment by szarnyasg** [comment]
This has been fixed indeed, the second experiments returns the correct results:

```
┌──────────────┐
│ count_star() │
│    int64     │
├──────────────┤
│        10000 │
└──────────────┘
```
